plugins/field-date/src/field_date.ts

```diff
@@ -0,0 +1,262 @@
+    const newDate = typeof newValue === 'string' ? new Date(newValue) : null;
+    if (!newDate || isNaN(newDate.getTime())) return null;
+
+    // NOTE: 'newValue' should be a valid date format here.
+    if (isISOFormat(newValue)) return newValue;
+
+    // Assume the time needs to be corrected.
+    return toLocalISOString(newDate);
+  }
+
+  /**
+   * Get the text to display on the block when the input hasn't spawned in.
+   *
+   * @returns The text to display on the block.
+   * @override
+   */
+  protected getText_(): string | null {
+    const value = this.getValue();
+    if (!value) return null;
+    // NOTE: There may be discrepancies between the text and the input based on
+    // browser. For example, 'en-US' would display the text '2/14/2020', then
+    // clicking in Safari on iOS would display 'Feb 14, 2020'.
+    return getLocaleDateString(value);
+  }
+
+  /**
+   * Renders the field. If the picker is shown make sure it has the current
+   * date selected.
+   */
+  protected render_() {
+    super.render_();
+  }
+
+  /**
+   * Shows the inline free-text editor on top of the text along with the date
+   * editor.
+   * @param e Optional mouse event that triggered the field to
+   *     open, or undefined if triggered programmatically.
+   * @param _quietInput Quiet input.
+   * @override
+   */
+  protected showEditor_(e?: Event, _quietInput?: boolean) {
+    if (this.sourceBlock_) {
+      // NOTE: Disable modal inputs to use default browser inputs on mobile.
+      this.sourceBlock_.workspace.options.modalInputs = false;
+    }
+
+    super.showEditor_(e, _quietInput);
+    this.showDropdown();
+  }
+
+  /**
+   * Updates the size of the field based on the text.
+   *
+   * @param margin margin to use when positioning the text element.
+   * @override
+   */
+  protected updateSize_(margin?: number) {
+    // Add margin so that the date input's datepicker icon doesn't clip with
+    // the text when sized for the date.
+    super.updateSize_((margin ?? 0) + 20);
+  }
+
+  /**
+   * Shows the datepicker.
+   */
+  private showDropdown(): void {
+    if (!this.htmlInput_) return;
+    Blockly.utils.dom.addClass(this.htmlInput_, 'blocklyDateInput');
+
+    // NOTE: HTMLInputElement.showPicker() is not available in earlier
+    // TypeScript versions (like 4.7.4), so casting to `any` to be compatible
+    // with dev scripts. Additionally, it's not available for date inputs for
+    // Safari. For browser compatibility of showPicker, see:
+    // https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/showPicker
+    /* eslint-disable @typescript-eslint/no-explicit-any */
+    (this.htmlInput_ as any).showPicker();
+    /* eslint-enable @typescript-eslint/no-explicit-any */
+  }
+
+  /**
+   * Create the html input and set it to type date.
+   *
+   * @returns The newly created date input editor.
+   */
+  protected widgetCreate_(): HTMLInputElement {
+    // NOTE: field_input should return HTMLInputElement for this.
```

Comment:
Can you file a bug to fix this in core? If you're busy I can instead :P

Reply:
Done! https://github.com/google/blockly/issues/7253

I created it as a feature request since it's not *technically* a bug, it could just have more type info.